File: src/main/java/skyblockclient/features/dungeons/Terminals.java
package skyblockclient.features.dungeons;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.inventory.GuiChest;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.init.Blocks;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.ContainerChest;
import net.minecraft.inventory.Slot;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.StringUtils;
import net.minecraftforge.client.event.GuiScreenEvent;
import net.minecraftforge.event.entity.player.ItemTooltipEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import skyblockclient.SkyblockClient;
import skyblockclient.events.GuiContainerEvent;
import skyblockclient.utils.ColorSlot;
import skyblockclient.utils.SkyblockCheck;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Terminals {

    private static final Minecraft mc = Minecraft.getMinecraft();
    private static final ArrayList<Slot> clickQueue = new ArrayList<>(28);
    private static TerminalType currentTerminal = TerminalType.NONE;
    private static long lastClickTime;
    private static int windowId;
    private static int windowClicks;
    private static boolean recalculate = false;

    @SubscribeEvent
    public void guiDraw(GuiScreenEvent.BackgroundDrawnEvent event) {
        if (!SkyblockCheck.inDungeons) return;
        if (event.gui instanceof GuiChest) {
            Container container = ((GuiChest) event.gui).inventorySlots;
            if (container instanceof ContainerChest) {
                List<Slot> invSlots = container.inventorySlots;
                if (currentTerminal == TerminalType.NONE) {
                    String chestName = ((ContainerChest) container).getLowerChestInventory().getDisplayName().getUnformattedText();
                    if (chestName.equals("Navigate the maze!")) {
                        currentTerminal = TerminalType.MAZE;
                    } else if (chestName.equals("Click in order!")) {
                        currentTerminal = TerminalType.NUMBERS;
                    } else if (chestName.equals("Correct all the panes!")) {
                        currentTerminal = TerminalType.CORRECTALL;
                    } else if (chestName.startsWith("What starts with:")) {
                        currentTerminal = TerminalType.LETTER;
                    } else if (chestName.startsWith("Select all the")) {
                        currentTerminal = TerminalType.COLOR;
                    }
                }
                if (currentTerminal != TerminalType.NONE) {
                    if (clickQueue.isEmpty() || recalculate) {
                        recalculate = getClicks((ContainerChest) container);
                    } else {
                        switch (currentTerminal) {
                            case MAZE:
                            case NUMBERS:
                            case CORRECTALL:
                                clickQueue.removeIf(slot -> invSlots.get(slot.slotNumber).getHasStack() &&
                                        invSlots.get(slot.slotNumber).getStack().getItemDamage() == 5);
                                break;
                            case LETTER:
                            case COLOR:
                                clickQueue.removeIf(slot -> invSlots.get(slot.slotNumber).getHasStack() &&
                                        invSlots.get(slot.slotNumber).getStack().isItemEnchanted());
                                break;
                        }
                    }
                    if (!clickQueue.isEmpty()) {
                        int size = invSlots.size();
                        switch (currentTerminal) {
                            case NUMBERS:
                                if (SkyblockClient.config.terminalHighlightNumbers) {
                                    new ColorSlot(clickQueue.get(0), size, SkyblockClient.config.terminalColorNumberFirst.getRGB());
                                    if (clickQueue.size() > 2) {
                                        new ColorSlot(clickQueue.get(2), size, SkyblockClient.config.terminalColorNumberThird.getRGB());
                                    }
                                    if (clickQueue.size() > 1) {
                                        new ColorSlot(clickQueue.get(1), size, SkyblockClient.config.terminalColorNumberSecond.getRGB());
                                    }
                                }
                                break;
                            case LETTER:
                                if (SkyblockClient.config.terminalHighlightLetter) {
                                    clickQueue.forEach(slot -> new ColorSlot(slot, size, SkyblockClient.config.terminalColorHighlight.getRGB()));
                                }
                                break;
                            case COLOR:
                                if (SkyblockClient.config.terminalHighlightColor) {
                                    clickQueue.forEach(slot -> new ColorSlot(slot, size, SkyblockClient.config.terminalColorHighlight.getRGB()));
                                }
                                break;
                        }
                        if (SkyblockClient.config.terminalAuto && System.currentTimeMillis() - lastClickTime > SkyblockClient.config.terminalClickDelay) {
                            switch (currentTerminal) {
                                case MAZE:
                                    if (SkyblockClient.config.terminalMaze) clickSlot(clickQueue.get(0));
                                    break;
                                case NUMBERS:
                                    if (SkyblockClient.config.terminalNumbers) clickSlot(clickQueue.get(0));
                                    break;
                                case CORRECTALL:
                                    if (SkyblockClient.config.terminalCorrectAll) clickSlot(clickQueue.get(0));
                                    break;
                                case LETTER:
                                    if (SkyblockClient.config.terminalLetter) clickSlot(clickQueue.get(0));
                                    break;
                                case COLOR:
                                    if (SkyblockClient.config.terminalColor) clickSlot(clickQueue.get(0));
                                    break;
                            }
                        }
                    }
                }
            }
        }
    }

    @SubscribeEvent
    public void drawSlot(GuiContainerEvent.DrawSlotEvent.Pre event) {
        if (!SkyblockClient.config.terminalHighlightNumbers) return;
        if (event.container instanceof ContainerChest && currentTerminal == TerminalType.NUMBERS) {
            Slot slot = event.slot;
            if (slot.inventory != mc.thePlayer.inventory && slot.getHasStack()) {
                ItemStack item = slot.getStack();
                if (item.getItem() == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.getItemDamage() == 14) {
                    GlStateManager.pushMatrix();
                    GlStateManager.disableLighting();
                    GlStateManager.disableDepth();
                    GlStateManager.disableBlend();
                    mc.fontRendererObj.drawStringWithShadow(String.valueOf(item.stackSize),
                            (float) (slot.xDisplayPosition + 9 - mc.fontRendererObj.getStringWidth(String.valueOf(item.stackSize)) / 2),
                            (float) (slot.yDisplayPosition + 4), 16777215);
                    GlStateManager.popMatrix();
                    event.setCanceled(true);
                }
            }
        }
    }

    @SubscribeEvent
    public void slotClick(GuiContainerEvent.SlotClickEvent event) {
        if (!SkyblockCheck.inDungeons || !SkyblockClient.config.terminalCustomClicks ||
                clickQueue.isEmpty() || !event.slot.getHasStack()) return;
        event.setCanceled(true);
        if (SkyblockClient.config.terminalBlockClicks) {
            switch (currentTerminal) {
                case MAZE:
                case NUMBERS:
                    if (clickQueue.get(0).slotNumber != event.slotId) return;
                    break;
                case CORRECTALL:
                    if (event.slot.getStack().getItemDamage() == 5) return;
                    break;
                case COLOR:
                case LETTER:
                    if (clickQueue.stream().noneMatch(slot -> slot.slotNumber == event.slotId)) return;
            }
        }
        clickSlot(event.slot);
        if (currentTerminal == TerminalType.CORRECTALL) {
            if (event.slot.getStack().getItemDamage() == 5 && SkyblockClient.config.terminalBlockClicks) {
                recalculate = true;
            }
        }
    }

    @SubscribeEvent
    public void tick(TickEvent.ClientTickEvent event) {
        if (!SkyblockCheck.inDungeons || event.phase != TickEvent.Phase.START) return;
        if (!(mc.currentScreen instanceof GuiChest)) {
            currentTerminal = TerminalType.NONE;
            clickQueue.clear();
            windowClicks = 0;
        }
    }

    @SubscribeEvent
    public void tooltip(ItemTooltipEvent event) {
        if (!SkyblockCheck.inDungeons || !SkyblockClient.config.terminalHideTooltip ||
                event.toolTip == null || currentTerminal == TerminalType.NONE) return;
        event.toolTip.clear();
    }

    private boolean getClicks(ContainerChest container) {
        List<Slot> invSlots = container.inventorySlots;
        String chestName = container.getLowerChestInventory().getDisplayName().getUnformattedText();
        clickQueue.clear();
        switch (currentTerminal) {
            case MAZE:
                int[] mazeSlotDirection = {-9, -1, 1, 9};
                boolean[] isStartSlot = new boolean[54];
                int endSlot = -1;
                boolean[] mazeVisited = new boolean[54];
                for (Slot slot : invSlots) {
                    if (slot.inventory == mc.thePlayer.inventory) continue;
                    ItemStack itemStack = slot.getStack();
                    if (itemStack == null) continue;
                    if (itemStack.getItem() == Item.getItemFromBlock(Blocks.stained_glass_pane)) {
                        if (itemStack.getItemDamage() == 5) {
                            isStartSlot[slot.slotNumber] = true;
                        } else if (itemStack.getItemDamage() == 14) {
                            endSlot = slot.slotNumber;
                        }
                    }
                }
                for (int j = 0; j < 54; j++) {
                    if (isStartSlot[j]) {
                        int startSlot = j;
                        boolean newSlotChosen;
                        while (startSlot != endSlot) {
                            newSlotChosen = false;
                            for (int i = 0; i < 4; i++) {
                                int slotNumber = startSlot + mazeSlotDirection[i];
                                if (slotNumber == endSlot) return false;
                                if (slotNumber < 0 || slotNumber > 53 || i == 1 && slotNumber % 9 == 0 || i == 2 && slotNumber % 9 == 8)
                                    continue;
                                if (mazeVisited[slotNumber]) continue;
                                ItemStack itemStack = invSlots.get(slotNumber).getStack();
                                if (itemStack == null) continue;
                                if (itemStack.getItem() == Item.getItemFromBlock(Blocks.stained_glass_pane) && itemStack.getItemDamage() == 0) {
                                    clickQueue.add(invSlots.get(slotNumber));
                                    startSlot = slotNumber;
                                    mazeVisited[slotNumber] = true;
                                    newSlotChosen = true;
                                    break;
                                }
                            }
                            if (!newSlotChosen) {
                                break;
                            }
                        }
                    }
                }
                return true;
            case NUMBERS:
                while (clickQueue.size() < 14) clickQueue.add(null);
                int min = 0;
                for (int i = 10; i <= 25; i++) {
                    if (i == 17 || i == 18) continue;
                    ItemStack item = invSlots.get(i).getStack();
                    if (item == null) continue;
                    if (item.getItem() == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.stackSize < 15) {
                        if (item.getItemDamage() == 14) {
                            clickQueue.set(item.stackSize - 1, invSlots.get(i));
                        } else if (item.getItemDamage() == 5) {
                            min = Math.max(0, item.stackSize - 1);
                        }
                    }
                }
                for (int i = min; i < 14; i++) {
                    if (clickQueue.get(i) == null) {
                        clickQueue.removeIf(Objects::isNull);
                        return true;
                    }
                }
                clickQueue.removeIf(Objects::isNull);
                break;
            case CORRECTALL:
                for (Slot slot : invSlots) {
                    if (slot.inventory == mc.thePlayer.inventory) continue;
                    if (slot.slotNumber < 9 || slot.slotNumber > 35 || slot.slotNumber % 9 <= 1 || slot.slotNumber % 9 >= 7)
                        continue;
                    ItemStack item = slot.getStack();
                    if (item == null) return true;
                    if (item.getItem() == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.getItemDamage() == 14) {
                        clickQueue.add(slot);
                    }
                }
                break;
            case LETTER:
                char letterNeeded = chestName.charAt(chestName.indexOf("'") + 1);
                if (letterNeeded != 0) {
                    for (Slot slot : invSlots) {
                        if (slot.inventory == mc.thePlayer.inventory) continue;
                        if (slot.slotNumber < 9 || slot.slotNumber > 44 || slot.slotNumber % 9 == 0 || slot.slotNumber % 9 == 8)
                            continue;
                        ItemStack item = slot.getStack();
                        if (item == null) return true;
                        if (item.isItemEnchanted()) continue;
                        if (StringUtils.stripControlCodes(item.getDisplayName()).charAt(0) == letterNeeded) {
                            clickQueue.add(slot);
                        }
                    }
                }
                break;
            case COLOR:
                String colorNeeded = null;
                for (EnumDyeColor color : EnumDyeColor.values()) {
                    String colorName = color.getName().replaceAll("_", " ").toUpperCase();
                    if (chestName.contains(colorName)) {
                        colorNeeded = color.getUnlocalizedName();
                        break;
                    }
                }
                if (colorNeeded != null) {
                    for (Slot slot : invSlots) {
                        if (slot.inventory == mc.thePlayer.inventory) continue;
                        if (slot.slotNumber < 9 || slot.slotNumber > 44 || slot.slotNumber % 9 == 0 || slot.slotNumber % 9 == 8)
                            continue;
                        ItemStack item = slot.getStack();
                        if (item == null) return true;
                        if (item.isItemEnchanted()) continue;
                        if (item.getUnlocalizedName().contains(colorNeeded)) {
                            clickQueue.add(slot);
                        }
                    }
                }
                break;
        }
        return false;
    }

    private void clickSlot(Slot slot) {
        if (windowClicks == 0) windowId = mc.thePlayer.openContainer.windowId;
        mc.playerController.windowClick(windowId + windowClicks, slot.slotNumber,
                SkyblockClient.config.terminalMiddleClick ? 2 : SkyblockClient.config.terminalShiftClick ? 1 : 0, 0, mc.thePlayer);
        lastClickTime = System.currentTimeMillis();
        if (SkyblockClient.config.terminalPingless) {
            windowClicks++;
            clickQueue.remove(slot);
        }
        if (clickQueue.size() == 1 && clickQueue.get(0).slotNumber == slot.slotNumber) {
            clickQueue.clear();
        }
    }

    private enum TerminalType {
        MAZE, NUMBERS, CORRECTALL, LETTER, COLOR, NONE
    }

}

File: src/main/kotlin/skyblockclient/features/dungeons/Terminals.kt
package skyblockclient.features.dungeons

import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.inventory.GuiChest
import net.minecraft.init.Blocks
import net.minecraft.inventory.ContainerChest
import net.minecraft.inventory.Slot
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.Item
import net.minecraft.util.StringUtils.stripControlCodes
import net.minecraftforge.client.event.GuiScreenEvent.BackgroundDrawnEvent
import net.minecraftforge.event.entity.player.ItemTooltipEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import skyblockclient.SkyblockClient.Companion.config
import skyblockclient.SkyblockClient.Companion.mc
import skyblockclient.events.GuiContainerEvent.DrawSlotEvent
import skyblockclient.events.GuiContainerEvent.SlotClickEvent
import skyblockclient.utils.Utils.equalsOneOf
import skyblockclient.utils.Utils.isFloor
import skyblockclient.utils.Utils.renderText
import kotlin.math.roundToInt

class Terminals {

    private val clickQueue = ArrayList<Slot>(28)
    private var currentTerminal = TerminalType.NONE
    private var lastClickTime: Long = 0
    private var recalculate = false
    private var shouldClick = false
    private var windowId = 0
    private var windowClicks = 0
    private var closestColor = 15

    @SubscribeEvent
    fun onGuiDraw(event: BackgroundDrawnEvent) {
        if (!isFloor(7) || event.gui !is GuiChest) return
        val container = (event.gui as GuiChest).inventorySlots
        if (container is ContainerChest) {
            if (currentTerminal == TerminalType.NONE) {
                val chestName = container.lowerChestInventory.displayName.unformattedText
                currentTerminal = when {
                    chestName == "Navigate the maze!" -> TerminalType.MAZE
                    chestName == "Click in order!" -> TerminalType.NUMBERS
                    chestName == "Correct all the panes!" -> TerminalType.CORRECTALL
                    chestName.startsWith("What starts with: '") -> TerminalType.LETTER
                    chestName.startsWith("Select all the") -> TerminalType.COLOR
                    chestName == "Click the button on time!" -> TerminalType.TIMING
                    chestName == "Change all to same color!" -> TerminalType.TOGGLECOLOR
                    else -> TerminalType.NONE
                }
                shouldClick = true
            }
            if (currentTerminal != TerminalType.NONE) {
                if (clickQueue.isEmpty() || recalculate) {
                    recalculate = getClicks(container)
                } else {
                    val invSlots = container.inventorySlots
                    shouldClick = when (currentTerminal) {
                        TerminalType.MAZE, TerminalType.NUMBERS, TerminalType.CORRECTALL -> clickQueue.removeIf {
                            invSlots[it.slotNumber].hasStack && invSlots[it.slotNumber].stack.itemDamage == 5
                        }
                        TerminalType.LETTER, TerminalType.COLOR -> clickQueue.removeIf {
                            invSlots[it.slotNumber].hasStack && invSlots[it.slotNumber].stack.isItemEnchanted
                        }
                        else -> true
                    } || shouldClick
                    if (clickQueue.isNotEmpty() && config.terminalAuto && (shouldClick || config.terminalPingless) &&
                        System.currentTimeMillis() - lastClickTime > config.terminalClickDelay
                    ) {
                        clickSlot(clickQueue[0])
                        shouldClick = false
                    }
                }
            }
            if (config.showTerminalInfo) {
                renderText("Terminal: " + currentTerminal.name, 20, 20)
                renderText("Clicks left: " + clickQueue.size, 20, 40)
                renderText("Window ID: $windowId, Pingless Clicks: $windowClicks", 20, 60)
            }
        }
    }

    @SubscribeEvent
    fun onDrawSlot(event: DrawSlotEvent) {
        if (!config.terminalHighlight || !isFloor(7) || event.gui !is GuiChest || event.slot.stack == null) return
        val x = event.slot.xDisplayPosition
        val y = event.slot.yDisplayPosition
        if (event.slot.inventory != mc.thePlayer.inventory) {
            when (currentTerminal) {
                TerminalType.NUMBERS -> {
                    (0..2).find {
                        clickQueue.size > it && event.slot.slotNumber == clickQueue[it].slotNumber
                    }?.let { Gui.drawRect(x, y, x + 16, y + 16, getColor(it)) }
                    if (event.slot.inventory != mc.thePlayer.inventory) {
                        val item = event.slot.stack
                        if (item.item == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.itemDamage == 14) {
                            renderText(
                                text = item.stackSize.toString(),
                                x = x + 9 - mc.fontRendererObj.getStringWidth(item.stackSize.toString()) / 2,
                                y = y + 4
                            )
                            event.isCanceled = true
                        }
                    }
                }
                TerminalType.LETTER, TerminalType.COLOR -> if (clickQueue.any { it.slotNumber == event.slot.slotNumber }) {
                    Gui.drawRect(x, y, x + 16, y + 16, config.terminalColorHighlight.rgb)
                }
                else -> {
                }
            }
        }
    }

    @SubscribeEvent
    fun onSlotClick(event: SlotClickEvent) {
        if (!config.terminalPingless && !config.terminalBlockClicks && !config.terminalMiddleClick ||
            !isFloor(7) || clickQueue.isEmpty() || event.slot == null
        ) return
        event.isCanceled = true
        if (config.terminalBlockClicks) {
            when (currentTerminal) {
                TerminalType.MAZE, TerminalType.NUMBERS -> if (clickQueue[0].slotNumber != event.slotId) return
                TerminalType.CORRECTALL -> if (event.slot?.stack?.itemDamage == 5) return
                TerminalType.COLOR, TerminalType.LETTER -> if (clickQueue.none { it.slotNumber == event.slotId }) return
                else -> {
                }
            }
        }
        clickSlot(event.slot!!)
    }

    @SubscribeEvent
    fun onTick(event: TickEvent.ClientTickEvent) {
        if (event.phase != TickEvent.Phase.START || !isFloor(7)) return
        if (mc.currentScreen is GuiChest) {
            if (config.terminalAutoNew && System.currentTimeMillis() - lastClickTime > config.terminalClickDelay) {
                val invSlots = mc.thePlayer.openContainer.inventorySlots
                when (currentTerminal) {
                    TerminalType.TIMING -> {
                        val currentRow = invSlots.indexOfFirst {
                            it.inventory != mc.thePlayer.inventory &&
                                    it.stack?.item == Item.getItemFromBlock(Blocks.stained_glass_pane) &&
                                    it.stack?.itemDamage == 10
                        }
                        if (currentRow == -1) return
                        val greenPaneIndex = invSlots.indexOfFirst {
                            it.inventory != mc.thePlayer.inventory &&
                                    it.stack?.item == Item.getItemFromBlock(Blocks.stained_glass_pane) &&
                                    it.stack?.itemDamage == 5
                        }
                        if (greenPaneIndex == -1) return
                        if (currentRow % 9 == greenPaneIndex % 9) {
                            mc.playerController.windowClick(
                                mc.thePlayer.openContainer.windowId,
                                (greenPaneIndex / 9) * 9 + 7,
                                if (config.terminalMiddleClick) 2 else 0,
                                1,
                                mc.thePlayer
                            )
                        }
                    }
                    TerminalType.TOGGLECOLOR -> {
                        if (closestColor == 15) {
                            closestColor = findClosestColor(invSlots.filter {
                                it.inventory != mc.thePlayer.inventory && it.stack?.item == Item.getItemFromBlock(
                                    Blocks.stained_glass_pane
                                ) && it.stack?.itemDamage != 15
                            })
                        } else {
                            val pane = invSlots.firstOrNull {
                                it.inventory != mc.thePlayer.inventory && it.stack?.item == Item.getItemFromBlock(
                                    Blocks.stained_glass_pane
                                ) && it.stack?.itemDamage != closestColor && it.stack?.itemDamage != 15
                            } ?: return
                            clickSlot(pane)
                        }
                    }
                    else -> return
                }
            }
        } else {
            currentTerminal = TerminalType.NONE
            clickQueue.clear()
            shouldClick = false
            closestColor = 15
        }
    }

    @SubscribeEvent
    fun onTooltip(event: ItemTooltipEvent) {
        if (!config.terminalHideTooltip || !isFloor(7) || currentTerminal == TerminalType.NONE || event.toolTip == null) return
        event.toolTip.clear()
    }

    private fun getColor(index: Int): Int {
        return when (index) {
            0 -> config.terminalColorNumberFirst.rgb
            1 -> config.terminalColorNumberSecond.rgb
            2 -> config.terminalColorNumberThird.rgb
            else -> 0xFFFFFF
        }
    }

    private fun getClicks(container: ContainerChest): Boolean {
        val invSlots = container.inventorySlots
        clickQueue.clear()
        windowClicks = 0
        when (currentTerminal) {
            TerminalType.MAZE -> {
                val startSlots = ArrayList<Slot>()
                var endSlot = -1
                invSlots.filter {
                    it.inventory != mc.thePlayer.inventory && it.stack?.item == Item.getItemFromBlock(Blocks.stained_glass_pane)
                }.forEach {
                    when (it.stack.itemDamage) {
                        5 -> startSlots.add(it)
                        14 -> endSlot = it.slotNumber
                    }
                }
                for (slot in startSlots) {
                    val mazeVisited = BooleanArray(54)
                    var start = slot.slotNumber
                    while (start != endSlot) {
                        var newSlotChosen = false
                        for (i in intArrayOf(-9, -1, 1, 9)) {
                            val next = start + i
                            if (next < 0 || next > 53 || i == -1 && start % 9 == 0 || i == 1 && start % 9 == 8) continue
                            if (mazeVisited[next]) continue
                            if (next == endSlot) return false
                            val itemStack = invSlots[next].stack ?: continue
                            if (itemStack.item == Item.getItemFromBlock(Blocks.stained_glass_pane) && itemStack.itemDamage == 0) {
                                clickQueue.add(invSlots[next])
                                start = next
                                mazeVisited[next] = true
                                newSlotChosen = true
                                break
                            }
                        }
                        if (!newSlotChosen) break
                    }
                }
                return true
            }
            TerminalType.NUMBERS -> {
                var min = 0
                val temp = arrayOfNulls<Slot>(14)
                for (i in 10..25) {
                    if (i == 17 || i == 18) continue
                    val item = invSlots[i].stack ?: return true
                    if (item.item == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.stackSize < 15) {
                        when (item.itemDamage) {
                            5 -> min = 0.coerceAtLeast(item.stackSize)
                            14 -> temp[item.stackSize - 1] = invSlots[i]
                        }
                    }
                }
                clickQueue.addAll(temp.filterNotNull())
                if (clickQueue.size != 14 - min) return true
            }
            TerminalType.CORRECTALL -> {
                for (slot in invSlots) {
                    if (slot.slotNumber !in 10..34 || slot.slotNumber % 9 <= 1 || slot.slotNumber % 9 >= 7) continue
                    val item = slot.stack ?: return true
                    if (item.item == Item.getItemFromBlock(Blocks.stained_glass_pane) && item.itemDamage == 14) {
                        clickQueue.add(slot)
                    }
                }
            }
            TerminalType.LETTER -> {
                val chestName = container.lowerChestInventory.displayName.unformattedText
                if (chestName.length > chestName.indexOf("'") + 1) {
                    val letterNeeded = chestName[chestName.indexOf("'") + 1]
                    for (slot in invSlots) {
                        if (slot.slotNumber !in 10..43 || (slot.slotNumber % 9).equalsOneOf(0, 8)) continue
                        val item = slot.stack ?: return true
                        if (!item.isItemEnchanted && stripControlCodes(item.displayName)[0] == letterNeeded) {
                            clickQueue.add(slot)
                        }
                    }
                }
            }
            TerminalType.COLOR -> {
                val colorNeeded = EnumDyeColor.values().find {
                    container.lowerChestInventory.displayName.unformattedText.contains(
                        it.getName().replace("_", " ").uppercase()
                    )
                }?.unlocalizedName ?: return false
                for (slot in invSlots) {
                    if (slot.slotNumber !in 10..43 || (slot.slotNumber % 9).equalsOneOf(0, 8)) continue
                    val item = slot.stack ?: return true
                    if (!item.isItemEnchanted && item.unlocalizedName.contains(colorNeeded)) {
                        clickQueue.add(slot)
                    }
                }
            }
            else -> {}
        }
        return false
    }

    private fun findClosestColor(panes: List<Slot>): Int {
        val mapping = mapOf(
            14 to 0,
            1 to 1,
            4 to 2,
            5 to 3,
            11 to 4
        )
        var sum = 0.0
        panes.forEach { sum += mapping[it.stack?.itemDamage] ?: 0 }
        val index = (sum / panes.size).roundToInt()
        mapping.entries.forEach { (key, value) ->
            if (value == index) return key
        }
        return 15
    }

    private fun clickSlot(slot: Slot) {
        if (windowClicks == 0) windowId = mc.thePlayer.openContainer.windowId
        mc.playerController.windowClick(
            windowId + windowClicks,
            slot.slotNumber,
            if (config.terminalMiddleClick) 2 else 0,
            0,
            mc.thePlayer
        )
        lastClickTime = System.currentTimeMillis()
        if (config.terminalPingless) {
            windowClicks++
            clickQueue.remove(slot)
        }
    }

    private enum class TerminalType {
        MAZE, NUMBERS, CORRECTALL, LETTER, COLOR, TIMING, TOGGLECOLOR, NONE
    }
}
